src/pmu.rs

```diff
@@ -0,0 +1,297 @@
+    /// 
+    /// # Errors
+    ///
+    /// * `BackupError::DataTooLarge` - returned if `user_data` cannot fit into backup registers
+    /// * `BackupError::DataMisaligned` - returned if `user_data` is not aligned to at least 4 bytes
+    ///
+    /// # Notes
+    ///
+    /// You can use `#[repr(align(4))]` to enforce a minimum alignment of 4 bytes for `user_data`
+    ///
+    fn restore_backup<UD>(&self, user_data: UD) -> Result<UD, BackupError>;
+
+    ///
+    /// Clears all backup registers by setting each to zero
+    ///
+    fn clear_backup(&self);
+
+    ///
+    /// Returns an enumified version of the Wakeup and Reset causes from the pmucause register
+    /// 
+    /// # Returns
+    /// 
+    /// * `Result<WakeupCause, CauseError>` - the cause enum is returned on success
+    /// 
+    /// # Errors
+    /// 
+    /// * `CauseError::InvalidCause` - returned if an unknown wakeup or reset cause is encountered
+    /// 
+    fn wakeup_cause(&self) -> Result<WakeupCause, CauseError>;
+}
+
+impl PMUExt for PMU {
+    fn load_default_programs(&self) {
+        unsafe {
+            for i in 0..8 {
+                self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                self.pmusleeppm[i].write(|w| w.bits(DEFAULT_SLEEP_PROGRAM[i]));
+
+                self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                self.pmuwakepm[i].write(|w| w.bits(DEFAULT_WAKE_PROGRAM[i]));
+            }
+        }
+    }
+
+    fn sleep(self, sleep_time: u32) {
+        unsafe {
+            let rtc = RTC::ptr();
+
+            // set interrupt source to RTC enabled, each pmu register needs key set before write
+            self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            self.pmuie.write(|w| w.rtc().set_bit().dwakeup().set_bit());
+            // set RTC clock scale to once per second for easy calculation
+            (*rtc).rtccfg.write(|w| w.enalways().set_bit().scale().bits(15));
+            // get current RTC clock value scaled
+            let rtc_now = (*rtc).rtcs.read().bits();
+            // set RTC clock comparator
+            (*rtc).rtccmp.write(|w| w.bits(rtc_now + sleep_time));
+            // go to sleep for sleep_time seconds, need to set pmukey here as well
+            self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            self.pmusleep.write(|w| w.sleep().set_bit());
+        }
+    }
+
+    fn store_backup<UD>(&self, user_data: UD) -> Result<UD, BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() {
+                return Err(BackupError::DataTooLarge);
+            }
+
+            if ud_size % 4 != 0 {
+                return Err(BackupError::DataMisaligned);
+            }
+
+            let ptr = &user_data as *const _;
+            let ptr_u32 = ptr as *const u32;
+            let sliced = core::slice::from_raw_parts(ptr_u32, ud_size);
```

Comment:
Needs `let reg_count = ud_size / 4;` as in `restore_backup` below

Reply:
nice find thanks! fixed